File: examples/sre-agent/python/pr_tools.py
"""Mock PR and approval handlers for the SRE response flow."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Mapping


def read_string(value: Any) -> str:
    return str(value).strip() if value is not None else ""


def read_positive_int(value: Any, fallback: int) -> int:
    if isinstance(value, bool) or not isinstance(value, (int, float)):
        return fallback
    if value < 1:
        return fallback
    return int(value)


@dataclass
class PullRequestRecord:
    number: int
    title: str
    body: str
    diff: str
    merged: bool = False
    approved: bool = False


@dataclass
class PrToolState:
    auto_approve: bool = False
    prs: list[PullRequestRecord] = field(default_factory=list)


def create_pr_tool_state(auto_approve: bool = False) -> PrToolState:
    return PrToolState(auto_approve=auto_approve)


def run_open_pull_request(
    state: PrToolState, args: Mapping[str, Any], _context: object = None
) -> dict[str, Any]:
    title = read_string(args.get("title"))
    body = read_string(args.get("body"))
    diff = read_string(args.get("diff"))

    if not title or not body or not diff:
        return {
            "opened": False,
            "pr_number": None,
            "url": None,
            "error": "title, body, and diff are required",
        }

    pr_number = len(state.prs) + 1
    state.prs.append(
        PullRequestRecord(number=pr_number, title=title, body=body, diff=diff)
    )
    return {
        "opened": True,
        "pr_number": pr_number,
        "url": f"mock://infra/pull/{pr_number}",
        "error": None,
    }


def run_request_approval(
    state: PrToolState, args: Mapping[str, Any], _context: object = None
) -> dict[str, Any]:
    summary = read_string(args.get("summary"))
    if not summary:
        return {
            "decision": "rejected",
            "summary": None,
            "error": "summary is required",
            "message": None,
        }

    if not state.prs:
        return {
            "decision": "rejected",
            "summary": None,
            "error": "open_pull_request must run first",
            "message": None,
        }

    latest_pr = state.prs[-1]
    if not state.auto_approve:
        return {
            "decision": "pending",
            "summary": None,
            "error": None,
            "message": (
                "Human approval required. Re-run with --auto-approve for a "
                "non-interactive demo."
            ),
        }

    latest_pr.approved = True
    return {
        "decision": "approved",
        "summary": summary,
        "error": None,
        "message": None,
    }


def run_merge_pull_request(
    state: PrToolState, args: Mapping[str, Any], _context: object = None
) -> dict[str, Any]:
    pr_number = read_positive_int(args.get("pr_number"), 0)
    pr = next((record for record in state.prs if record.number == pr_number), None)

    if pr is None:
        return {
            "merged": False,
            "pr_number": None,
            "error": f"PR #{pr_number} not found",
        }

    if not pr.approved:
        return {
            "merged": False,
            "pr_number": None,
            "error": "PR is not approved. Call request_approval first.",
        }

    pr.merged = True
    return {"merged": True, "pr_number": pr.number, "error": None}

File: examples/sre-agent/ts/src/pr-tools.ts
import type { SDKJsonValue } from "@cursor/sdk";
import { readPositiveInt, readString } from "./tools.js";

export type PullRequestRecord = {
  number: number;
  title: string;
  body: string;
  diff: string;
  merged: boolean;
  approved: boolean;
};

export type PrToolState = {
  prs: PullRequestRecord[];
  autoApprove: boolean;
};

export function createPrToolState(autoApprove = false): PrToolState {
  return { prs: [], autoApprove };
}

export function openPullRequest(
  state: PrToolState,
  args: {
    title?: SDKJsonValue;
    body?: SDKJsonValue;
    diff?: SDKJsonValue;
  }
) {
  const title = readString(args.title);
  const body = readString(args.body);
  const diff = readString(args.diff);

  if (!title || !body || !diff) {
    return {
      opened: false,
      pr_number: null,
      url: null,
      error: "title, body, and diff are required"
    };
  }

  const prNumber = state.prs.length + 1;
  state.prs.push({
    number: prNumber,
    title,
    body,
    diff,
    merged: false,
    approved: false
  });

  return {
    opened: true,
    pr_number: prNumber,
    url: `mock://infra/pull/${prNumber}`,
    error: null
  };
}

export function requestApproval(
  state: PrToolState,
  args: { summary?: SDKJsonValue }
) {
  const summary = readString(args.summary);

  if (!summary) {
    return {
      decision: "rejected",
      summary: null,
      error: "summary is required",
      message: null
    };
  }

  const latestPr = state.prs.at(-1);
  if (!latestPr) {
    return {
      decision: "rejected",
      summary: null,
      error: "open_pull_request must run first",
      message: null
    };
  }

  if (!state.autoApprove) {
    return {
      decision: "pending",
      summary: null,
      error: null,
      message:
        "Human approval required. Re-run with --auto-approve for a non-interactive demo."
    };
  }

  latestPr.approved = true;
  return { decision: "approved", summary, error: null, message: null };
}

export function mergePullRequest(
  state: PrToolState,
  args: { pr_number?: SDKJsonValue }
) {
  const prNumber = readPositiveInt(args.pr_number, 0);
  const pr = state.prs.find((record) => record.number === prNumber);

  if (!pr) {
    return {
      merged: false,
      pr_number: null,
      error: `PR #${prNumber} not found`
    };
  }

  if (!pr.approved) {
    return {
      merged: false,
      pr_number: null,
      error: "PR is not approved. Call request_approval first."
    };
  }

  pr.merged = true;
  return { merged: true, pr_number: pr.number, error: null };
}
